x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/ManyShardsIT.java

```diff
@@ -283,17 +282,7 @@ public void testCancelUnnecessaryRequests() {
         try {
             var result = safeExecute(client(coordinatingNode), EsqlQueryAction.INSTANCE, query);
             assertThat(Iterables.size(result.rows()), equalTo(1L));
-            assertThat(exchanges.get(), lessThanOrEqualTo(1));// 0 if result is populated from coordinating node
-        } catch (AssertionError e) {
-            client().admin().indices().stats(new IndicesStatsRequest()).actionGet().asMap().forEach((shard, stats) -> {
-                logger.info(
-                    "Shard {} node {} status {} docs {}",
-                    shard.shardId(),
-                    shard.currentNodeId(),
-                    shard.state(),
-                    stats.getStats().getDocs().getCount()
-                );
-            });
+            assertThat(exchanges.get(), lessThanOrEqualTo(2));
```

Comment:
Once above was fixed I noticed that 1-8% depending on the load of the times the test was failing.

Since processing pages and sending more requests happens on different threads of different threadpools, we can not guarantee that incoming page is processed before the next request is sent.

It looks like in practice we accumulate enough results (1 row) before the third request is sent.
This still shows that we do not send request to every node (there are at minimum 3 data nodes in this test).